File: app/services/ports.py
import threading
from typing import Tuple, Optional, Dict
from ..core.config import cfg

class PortAllocator:
    def __init__(self):
        self._lock = threading.RLock()
        self._host_min, self._host_max = self._parse(cfg.PORT_RANGE_HOST)
        self._http_min, self._http_max = self._parse(cfg.ACE_HTTP_RANGE)
        self._https_min, self._https_max = self._parse(cfg.ACE_HTTPS_RANGE)
        self._host_next = self._host_min
        self._http_next = self._http_min
        self._https_next = self._https_min
        self._used_host: set[int] = set()
        self._used_http: set[int] = set()
        self._used_https: set[int] = set()
        
        # Gluetun-specific port allocation starting from 19000
        self._gluetun_port_base = 19000
        self._gluetun_min = max(self._gluetun_port_base, self._host_min)
        self._gluetun_max = self._host_max
        self._gluetun_next = self._gluetun_min
        self._used_gluetun_ports: set[int] = set()
        
        # VPN-specific port ranges for redundant mode
        # Maps VPN container name to (min_port, max_port, next_port, used_ports_set)
        self._vpn_port_ranges: Dict[str, Tuple[int, int, int, set[int]]] = {}
        self._init_vpn_port_ranges()

    def _parse(self, s: str) -> Tuple[int, int]:
        a, b = s.split("-")
        return int(a), int(b)
    
    def _init_vpn_port_ranges(self):
        """Initialize VPN-specific port ranges for redundant mode."""
        import logging
        logger = logging.getLogger(__name__)
        
        # Set up port ranges for each VPN if configured
        if cfg.GLUETUN_CONTAINER_NAME and cfg.GLUETUN_PORT_RANGE_1:
            try:
                min_port, max_port = self._parse(cfg.GLUETUN_PORT_RANGE_1)
                self._vpn_port_ranges[cfg.GLUETUN_CONTAINER_NAME] = (min_port, max_port, min_port, set())
                logger.info(f"VPN port range for {cfg.GLUETUN_CONTAINER_NAME}: {min_port}-{max_port}")
            except (ValueError, AttributeError) as e:
                logger.error(f"Invalid GLUETUN_PORT_RANGE_1 format '{cfg.GLUETUN_PORT_RANGE_1}': {e}. Expected format: 'min-max'")
        
        if cfg.GLUETUN_CONTAINER_NAME_2 and cfg.GLUETUN_PORT_RANGE_2:
            try:
                min_port, max_port = self._parse(cfg.GLUETUN_PORT_RANGE_2)
                self._vpn_port_ranges[cfg.GLUETUN_CONTAINER_NAME_2] = (min_port, max_port, min_port, set())
                logger.info(f"VPN port range for {cfg.GLUETUN_CONTAINER_NAME_2}: {min_port}-{max_port}")
            except (ValueError, AttributeError) as e:
                logger.error(f"Invalid GLUETUN_PORT_RANGE_2 format '{cfg.GLUETUN_PORT_RANGE_2}': {e}. Expected format: 'min-max'")

    def _next_in(self, cur: int, lo: int, hi: int, used: set[int]) -> int:
        p = cur
        for _ in range(hi - lo + 1):
            if p > hi:
                p = lo
            if p not in used:
                return p
            p += 1
        raise RuntimeError("no free ports in range")

    def alloc_host(self) -> int:
        with self._lock:
            p = self._next_in(self._host_next, self._host_min, self._host_max, self._used_host)
            self._used_host.add(p)
            self._host_next = p + 1
            return p

    def alloc_http(self) -> int:
        with self._lock:
            p = self._next_in(self._http_next, self._http_min, self._http_max, self._used_http)
            self._used_http.add(p)
            self._http_next = p + 1
            return p

    def alloc_https(self, avoid: Optional[int] = None) -> int:
        with self._lock:
            while True:
                p = self._next_in(self._https_next, self._https_min, self._https_max, self._used_https)
                if avoid is None or p != avoid:
                    self._used_https.add(p)
                    self._https_next = p + 1
                    return p
                self._https_next = p + 1

    def reserve_host(self, p: int):
        with self._lock:
            self._used_host.add(p)

    def reserve_http(self, p: int):
        with self._lock:
            self._used_http.add(p)

    def reserve_https(self, p: int):
        with self._lock:
            self._used_https.add(p)

    def free_host(self, p: Optional[int]):
        if p is None: return
        with self._lock:
            self._used_host.discard(p)

    def free_http(self, p: Optional[int]):
        if p is None: return
        with self._lock:
            self._used_http.discard(p)

    def free_https(self, p: Optional[int]):
        if p is None: return
        with self._lock:
            self._used_https.discard(p)

    def allocate_engine_ports(
        self,
        *,
        use_gluetun: bool,
        vpn_container: Optional[str],
        requested_host_port: Optional[int],
        user_http_port: Optional[int],
        user_https_port: Optional[int],
        user_api_port: Optional[int],
        map_https: bool,
    ) -> Dict[str, Optional[int]]:
        """Atomically reserve all ports required for a new engine.

        Returns a dictionary with container and host ports.
        If any allocation fails, all partial reservations are rolled back.
        """

        with self._lock:
            reservations: list[tuple[str, int, Optional[str]]] = []

            def _track(kind: str, port: int, vpn: Optional[str] = None):
                reservations.append((kind, port, vpn))

            def _rollback():
                for kind, port, vpn in reversed(reservations):
                    if kind == "host":
                        self.free_host(port)
                    elif kind == "http":
                        self.free_http(port)
                    elif kind == "https":
                        self.free_https(port)
                    elif kind == "gluetun":
                        self.free_gluetun_port(port, vpn)

            try:
                if user_http_port is not None:
                    c_http = user_http_port
                    host_http = requested_host_port or user_http_port
                    if use_gluetun:
                        self.reserve_gluetun_port(c_http, vpn_container)
                        _track("gluetun", c_http, vpn_container)
                    else:
                        self.reserve_http(c_http)
                        _track("http", c_http)
                else:
                    if use_gluetun:
                        host_http = self.alloc_gluetun_port(vpn_container)
                        _track("gluetun", host_http, vpn_container)
                        c_http = host_http
                    else:
                        host_http = requested_host_port or self.alloc_host()
                        if requested_host_port is None:
                            _track("host", host_http)
                        c_http = host_http
                        self.reserve_http(c_http)
                        _track("http", c_http)

                if user_https_port is not None:
                    c_https = user_https_port
                    self.reserve_https(c_https)
                    _track("https", c_https)
                else:
                    c_https = self.alloc_https(avoid=c_http)
                    _track("https", c_https)

                if user_api_port is not None:
                    c_api = user_api_port
                    host_api = user_api_port
                    if use_gluetun:
                        self.reserve_gluetun_port(host_api, vpn_container)
                        _track("gluetun", host_api, vpn_container)
                    else:
                        self.reserve_host(host_api)
                        _track("host", host_api)
                else:
                    if use_gluetun:
                        host_api = self.alloc_gluetun_port(vpn_container)
                        _track("gluetun", host_api, vpn_container)
                    else:
                        host_api = self.alloc_host()
                        _track("host", host_api)
                    c_api = host_api

                host_https = None
                if not use_gluetun and map_https:
                    host_https = self.alloc_host()
                    _track("host", host_https)

                return {
                    "container_http_port": c_http,
                    "container_https_port": c_https,
                    "container_api_port": c_api,
                    "host_http_port": host_http,
                    "host_api_port": host_api,
                    "host_https_port": host_https,
                }
            except Exception:
                _rollback()
                raise

    def alloc_gluetun_port(self, vpn_container: Optional[str] = None) -> int:
        """
        Allocate a port for Gluetun from the appropriate range.
        
        Args:
            vpn_container: VPN container name. If provided and VPN-specific ranges are configured,
                          allocates from that VPN's range. Otherwise uses global allocation.
        
        Returns:
            Allocated port number
        """
        with self._lock:
            # If VPN-specific range is configured, use it
            if vpn_container and vpn_container in self._vpn_port_ranges:
                min_port, max_port, next_port, used_ports = self._vpn_port_ranges[vpn_container]
                
                # Find next available port in this VPN's range
                port = self._next_in(next_port, min_port, max_port, used_ports)
                used_ports.add(port)
                
                # Update next_port for this VPN
                self._vpn_port_ranges[vpn_container] = (min_port, max_port, port + 1, used_ports)
                return port
            
            # Fallback to global allocation (backwards compatibility)
            if self._gluetun_min > self._gluetun_max:
                raise RuntimeError("Invalid Gluetun fallback port range")

            port = self._next_in(self._gluetun_next, self._gluetun_min, self._gluetun_max, self._used_gluetun_ports)
            self._used_gluetun_ports.add(port)
            self._gluetun_next = port + 1
            return port

    def reserve_gluetun_port(self, p: int, vpn_container: Optional[str] = None):
        """Reserve a specific Gluetun port."""
        with self._lock:
            # If VPN-specific range is configured, reserve in that range
            if vpn_container and vpn_container in self._vpn_port_ranges:
                min_port, max_port, next_port, used_ports = self._vpn_port_ranges[vpn_container]
                used_ports.add(p)
                self._vpn_port_ranges[vpn_container] = (min_port, max_port, next_port, used_ports)
            else:
                # Fallback to global reservation
                self._used_gluetun_ports.add(p)

    def free_gluetun_port(self, p: Optional[int], vpn_container: Optional[str] = None):
        """Free a Gluetun port."""
        if p is None: return
        with self._lock:
            # If VPN-specific range is configured, free from that range
            if vpn_container and vpn_container in self._vpn_port_ranges:
                min_port, max_port, next_port, used_ports = self._vpn_port_ranges[vpn_container]
                used_ports.discard(p)
                self._vpn_port_ranges[vpn_container] = (min_port, max_port, next_port, used_ports)
            else:
                # Fallback to global freeing
                self._used_gluetun_ports.discard(p)
    
    def clear_all_allocations(self):
        """Clear all port allocations. Used during cleanup to reset state."""
        with self._lock:
            self._used_host.clear()
            self._used_http.clear()
            self._used_https.clear()
            self._used_gluetun_ports.clear()
            self._gluetun_next = self._gluetun_min
            
            # Clear VPN-specific port allocations
            for vpn_name in self._vpn_port_ranges:
                min_port, max_port, next_port, used_ports = self._vpn_port_ranges[vpn_name]
                used_ports.clear()
                self._vpn_port_ranges[vpn_name] = (min_port, max_port, min_port, used_ports)

alloc = PortAllocator()

File: app/orchestrator/internal/controlplane/engine/ports.go
package engine

import (
	"fmt"
	"strings"
	"sync"

	"github.com/acestream/acestream/internal/config"
)

// PortAllocator is the exact Go equivalent of the Python PortAllocator.
// All operations are safe for concurrent use.
type PortAllocator struct {
	mu sync.Mutex

	hostMin, hostMax int
	httpMin, httpMax int
	httpsMin, httpsMax int

	hostNext  int
	httpNext  int
	httpsNext int

	usedHost  map[int]struct{}
	usedHTTP  map[int]struct{}
	usedHTTPS map[int]struct{}

	// Gluetun global fallback range (starts at 19000, capped to host range)
	gluetunMin, gluetunMax int
	gluetunNext            int
	usedGluetun            map[int]struct{}

	// Per-VPN port ranges from GLUETUN_PORT_RANGE_1/2 env vars
	vpnRangeSlots       []portRange          // ordered slot list
	vpnRangeAssignments map[string]int        // vpnContainer -> slot index
	vpnPortRanges       map[string]*vpnRange  // vpnContainer -> range state

	// Per-VPN internal P2P ports (starting at 62062)
	vpnP2PUsed map[string]map[int]struct{}
}

type portRange struct{ min, max int }

type vpnRange struct {
	min, max int
	next     int
	used     map[int]struct{}
}

var Alloc = newPortAllocator()

func newPortAllocator() *PortAllocator {
	cfg := config.C.Load()

	a := &PortAllocator{
		hostMin: cfg.PortRangeHost.Min,
		hostMax: cfg.PortRangeHost.Max,
		httpMin: cfg.ACEHTTPRange.Min,
		httpMax: cfg.ACEHTTPRange.Max,
		httpsMin: cfg.ACEHTTPSRange.Min,
		httpsMax: cfg.ACEHTTPSRange.Max,

		usedHost:  make(map[int]struct{}),
		usedHTTP:  make(map[int]struct{}),
		usedHTTPS: make(map[int]struct{}),
		usedGluetun: make(map[int]struct{}),

		vpnRangeSlots:       nil,
		vpnRangeAssignments: make(map[string]int),
		vpnPortRanges:       make(map[string]*vpnRange),
		vpnP2PUsed:          make(map[string]map[int]struct{}),
	}

	a.hostNext = a.hostMin
	a.httpNext = a.httpMin
	a.httpsNext = a.httpsMin

	gluetunBase := 19000
	a.gluetunMin = max(gluetunBase, a.hostMin)
	a.gluetunMax = a.hostMax
	a.gluetunNext = a.gluetunMin

	// Parse optional VPN-specific port ranges
	for _, raw := range []string{cfg.GluetunPortRange1, cfg.GluetunPortRange2} {
		if raw == "" {
			continue
		}
		parts := strings.SplitN(raw, "-", 2)
		if len(parts) != 2 {
			continue
		}
		var lo, hi int
		fmt.Sscanf(parts[0], "%d", &lo)
		fmt.Sscanf(parts[1], "%d", &hi)
		if lo > 0 && hi >= lo {
			a.vpnRangeSlots = append(a.vpnRangeSlots, portRange{lo, hi})
		}
	}

	return a
}



func (a *PortAllocator) nextIn(cur, lo, hi int, used map[int]struct{}) (int, error) {
	p := cur
	for range hi - lo + 1 {
		if p > hi {
			p = lo
		}
		if _, inUse := used[p]; !inUse {
			return p, nil
		}
		p++
	}
	return 0, fmt.Errorf("no free ports in range %d-%d", lo, hi)
}

func (a *PortAllocator) AllocHost() (int, error) {
	a.mu.Lock()
	defer a.mu.Unlock()
	p, err := a.nextIn(a.hostNext, a.hostMin, a.hostMax, a.usedHost)
	if err != nil {
		return 0, err
	}
	a.usedHost[p] = struct{}{}
	a.hostNext = p + 1
	return p, nil
}

func (a *PortAllocator) AllocHTTP() (int, error) {
	a.mu.Lock()
	defer a.mu.Unlock()
	p, err := a.nextIn(a.httpNext, a.httpMin, a.httpMax, a.usedHTTP)
	if err != nil {
		return 0, err
	}
	a.usedHTTP[p] = struct{}{}
	a.httpNext = p + 1
	return p, nil
}

func (a *PortAllocator) AllocHTTPS(avoid int) (int, error) {
	a.mu.Lock()
	defer a.mu.Unlock()
	for {
		p, err := a.nextIn(a.httpsNext, a.httpsMin, a.httpsMax, a.usedHTTPS)
		if err != nil {
			return 0, err
		}
		if avoid == 0 || p != avoid {
			a.usedHTTPS[p] = struct{}{}
			a.httpsNext = p + 1
			return p, nil
		}
		a.httpsNext = p + 1
	}
}

func (a *PortAllocator) ReserveHost(p int) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.usedHost[p] = struct{}{}
}

func (a *PortAllocator) ReserveHTTP(p int) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.usedHTTP[p] = struct{}{}
}

func (a *PortAllocator) ReserveHTTPS(p int) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.usedHTTPS[p] = struct{}{}
}

func (a *PortAllocator) FreeHost(p int) {
	if p == 0 {
		return
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	delete(a.usedHost, p)
}

func (a *PortAllocator) FreeHTTP(p int) {
	if p == 0 {
		return
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	delete(a.usedHTTP, p)
}

func (a *PortAllocator) FreeHTTPS(p int) {
	if p == 0 {
		return
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	delete(a.usedHTTPS, p)
}

// ensureVPNRangeAssigned assigns a port-range slot to a VPN on first use.
// Must be called with a.mu held.
func (a *PortAllocator) ensureVPNRangeAssigned(vpn string) {
	if vpn == "" || len(a.vpnRangeSlots) == 0 {
		return
	}
	if _, ok := a.vpnPortRanges[vpn]; ok {
		return
	}
	// Find first unassigned slot
	used := make(map[int]struct{}, len(a.vpnRangeAssignments))
	for _, idx := range a.vpnRangeAssignments {
		used[idx] = struct{}{}
	}
	for idx, slot := range a.vpnRangeSlots {
		if _, taken := used[idx]; !taken {
			a.vpnRangeAssignments[vpn] = idx
			a.vpnPortRanges[vpn] = &vpnRange{
				min:  slot.min,
				max:  slot.max,
				next: slot.min,
				used: make(map[int]struct{}),
			}
			return
		}
	}
}

func (a *PortAllocator) AllocGluetunPort(vpn string) (int, error) {
	a.mu.Lock()
	defer a.mu.Unlock()

	a.ensureVPNRangeAssigned(vpn)

	if vr, ok := a.vpnPortRanges[vpn]; ok {
		p, err := a.nextIn(vr.next, vr.min, vr.max, vr.used)
		if err != nil {
			return 0, err
		}
		vr.used[p] = struct{}{}
		vr.next = p + 1
		return p, nil
	}

	// Global fallback
	p, err := a.nextIn(a.gluetunNext, a.gluetunMin, a.gluetunMax, a.usedGluetun)
	if err != nil {
		return 0, err
	}
	a.usedGluetun[p] = struct{}{}
	a.gluetunNext = p + 1
	return p, nil
}

func (a *PortAllocator) ReserveGluetunPort(p int, vpn string) {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.ensureVPNRangeAssigned(vpn)
	if vr, ok := a.vpnPortRanges[vpn]; ok {
		vr.used[p] = struct{}{}
	} else {
		a.usedGluetun[p] = struct{}{}
	}
}

func (a *PortAllocator) FreeGluetunPort(p int, vpn string) {
	if p == 0 {
		return
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	if vr, ok := a.vpnPortRanges[vpn]; ok {
		delete(vr.used, p)
	} else {
		delete(a.usedGluetun, p)
	}
}

// PortsInfo is the result of AllocateEnginePorts.
type PortsInfo struct {
	ContainerHTTPPort  int
	ContainerHTTPSPort int
	ContainerAPIPort   int
	HostHTTPPort       int
	HostAPIPort        int
	HostHTTPSPort      int // 0 if not mapped
}

// AllocateEnginePorts atomically reserves all ports for a new engine.
// Mirrors Python's PortAllocator.allocate_engine_ports exactly.
func (a *PortAllocator) AllocateEnginePorts(
	useGluetun bool,
	vpnContainer string,
	requestedHostPort int,
	userHTTPPort, userHTTPSPort, userAPIPort int,
	mapHTTPS bool,
) (PortsInfo, error) {
	a.mu.Lock()
	defer a.mu.Unlock()

	type reservation struct {
		kind string
		port int
		vpn  string
	}
	var reservations []reservation

	rollback := func() {
		for i := len(reservations) - 1; i >= 0; i-- {
			r := reservations[i]
			switch r.kind {
			case "host":
				delete(a.usedHost, r.port)
			case "http":
				delete(a.usedHTTP, r.port)
			case "https":
				delete(a.usedHTTPS, r.port)
			case "gluetun":
				if vr, ok := a.vpnPortRanges[r.vpn]; ok {
					delete(vr.used, r.port)
				} else {
					delete(a.usedGluetun, r.port)
				}
			}
		}
	}

	track := func(kind string, port int, vpn string) {
		reservations = append(reservations, reservation{kind, port, vpn})
	}

	a.ensureVPNRangeAssigned(vpnContainer)

	allocGluetun := func() (int, error) {
		if vr, ok := a.vpnPortRanges[vpnContainer]; ok {
			p, err := a.nextIn(vr.next, vr.min, vr.max, vr.used)
			if err != nil {
				return 0, err
			}
			vr.used[p] = struct{}{}
			vr.next = p + 1
			return p, nil
		}
		p, err := a.nextIn(a.gluetunNext, a.gluetunMin, a.gluetunMax, a.usedGluetun)
		if err != nil {
			return 0, err
		}
		a.usedGluetun[p] = struct{}{}
		a.gluetunNext = p + 1
		return p, nil
	}

	var info PortsInfo
	var err error

	// HTTP port
	if userHTTPPort != 0 {
		info.ContainerHTTPPort = userHTTPPort
		info.HostHTTPPort = requestedHostPort
		if info.HostHTTPPort == 0 {
			info.HostHTTPPort = userHTTPPort
		}
		if useGluetun {
			if vr, ok := a.vpnPortRanges[vpnContainer]; ok {
				vr.used[info.ContainerHTTPPort] = struct{}{}
			} else {
				a.usedGluetun[info.ContainerHTTPPort] = struct{}{}
			}
			track("gluetun", info.ContainerHTTPPort, vpnContainer)
		} else {
			a.usedHTTP[info.ContainerHTTPPort] = struct{}{}
			track("http", info.ContainerHTTPPort, "")
		}
	} else {
		if useGluetun {
			info.HostHTTPPort, err = allocGluetun()
			if err != nil {
				rollback()
				return PortsInfo{}, err
			}
			track("gluetun", info.HostHTTPPort, vpnContainer)
			info.ContainerHTTPPort = info.HostHTTPPort
		} else {
			if requestedHostPort != 0 {
				info.HostHTTPPort = requestedHostPort
			} else {
				info.HostHTTPPort, err = a.nextIn(a.hostNext, a.hostMin, a.hostMax, a.usedHost)
				if err != nil {
					rollback()
					return PortsInfo{}, err
				}
				a.usedHost[info.HostHTTPPort] = struct{}{}
				a.hostNext = info.HostHTTPPort + 1
				track("host", info.HostHTTPPort, "")
			}
			info.ContainerHTTPPort = info.HostHTTPPort
			a.usedHTTP[info.ContainerHTTPPort] = struct{}{}
			track("http", info.ContainerHTTPPort, "")
		}
	}

	// HTTPS port
	if userHTTPSPort != 0 {
		info.ContainerHTTPSPort = userHTTPSPort
		a.usedHTTPS[info.ContainerHTTPSPort] = struct{}{}
		track("https", info.ContainerHTTPSPort, "")
	} else {
		for {
			p, err2 := a.nextIn(a.httpsNext, a.httpsMin, a.httpsMax, a.usedHTTPS)
			if err2 != nil {
				rollback()
				return PortsInfo{}, err2
			}
			if p != info.ContainerHTTPPort {
				info.ContainerHTTPSPort = p
				a.usedHTTPS[p] = struct{}{}
				a.httpsNext = p + 1
				track("https", p, "")
				break
			}
			a.httpsNext = p + 1
		}
	}

	// API port
	if userAPIPort != 0 {
		info.ContainerAPIPort = userAPIPort
		info.HostAPIPort = userAPIPort
		if useGluetun {
			if vr, ok := a.vpnPortRanges[vpnContainer]; ok {
				vr.used[info.HostAPIPort] = struct{}{}
			} else {
				a.usedGluetun[info.HostAPIPort] = struct{}{}
			}
			track("gluetun", info.HostAPIPort, vpnContainer)
		} else {
			a.usedHost[info.HostAPIPort] = struct{}{}
			track("host", info.HostAPIPort, "")
		}
	} else {
		if useGluetun {
			info.HostAPIPort, err = allocGluetun()
			if err != nil {
				rollback()
				return PortsInfo{}, err
			}
			track("gluetun", info.HostAPIPort, vpnContainer)
		} else {
			info.HostAPIPort, err = a.nextIn(a.hostNext, a.hostMin, a.hostMax, a.usedHost)
			if err != nil {
				rollback()
				return PortsInfo{}, err
			}
			a.usedHost[info.HostAPIPort] = struct{}{}
			a.hostNext = info.HostAPIPort + 1
			track("host", info.HostAPIPort, "")
		}
		info.ContainerAPIPort = info.HostAPIPort
	}

	// Optional host-side HTTPS mapping
	if !useGluetun && mapHTTPS {
		info.HostHTTPSPort, err = a.nextIn(a.hostNext, a.hostMin, a.hostMax, a.usedHost)
		if err != nil {
			rollback()
			return PortsInfo{}, err
		}
		a.usedHost[info.HostHTTPSPort] = struct{}{}
		a.hostNext = info.HostHTTPSPort + 1
		track("host", info.HostHTTPSPort, "")
	}

	return info, nil
}

// ReserveFromLabels marks ports stored in container labels as in-use.
// Called on engine start events so allocator state stays consistent with running containers.
func (a *PortAllocator) ReserveFromLabels(labels map[string]string) {
	vpn := labels["acestream.vpn_container"]
	hp := parseInt(labels["host.http_port"])
	ap := parseInt(labels["host.api_port"])
	hsp := parseInt(labels["host.https_port"])
	cp := parseInt(labels["acestream.http_port"])
	sp := parseInt(labels["acestream.https_port"])

	a.mu.Lock()
	defer a.mu.Unlock()
	if hp != 0 {
		a.usedHost[hp] = struct{}{}
	}
	if cp != 0 {
		a.usedHTTP[cp] = struct{}{}
	}
	if sp != 0 {
		a.usedHTTPS[sp] = struct{}{}
	}
	if hsp != 0 {
		a.usedHost[hsp] = struct{}{}
	}
	if vpn != "" {
		// Reserve in per-VPN slot range when available, otherwise global gluetun pool.
		// Mirrors AllocateEnginePorts / ReserveGluetunPort logic (inline to avoid re-locking).
		a.ensureVPNRangeAssigned(vpn)
		reserveVPN := func(p int) {
			if p == 0 {
				return
			}
			if vr, ok := a.vpnPortRanges[vpn]; ok {
				vr.used[p] = struct{}{}
			} else {
				a.usedGluetun[p] = struct{}{}
			}
		}
		reserveVPN(hp)
		if ap != hp {
			reserveVPN(ap)
		}
	} else if ap != 0 && ap != hp {
		a.usedHost[ap] = struct{}{}
	}
}

// ReleaseFromLabels frees all ports stored in container labels (called on container die).
func (a *PortAllocator) ReleaseFromLabels(labels map[string]string) {
	vpn := labels["acestream.vpn_container"]
	hp := parseInt(labels["host.http_port"])
	ap := parseInt(labels["host.api_port"])
	hsp := parseInt(labels["host.https_port"])
	cp := parseInt(labels["acestream.http_port"])
	sp := parseInt(labels["acestream.https_port"])

	a.FreeHost(hp)
	a.FreeHTTP(cp)
	a.FreeHTTPS(sp)
	if hsp != 0 {
		a.FreeHost(hsp)
	}
	if vpn != "" {
		a.FreeGluetunPort(hp, vpn)
		if ap != 0 && ap != hp {
			a.FreeGluetunPort(ap, vpn)
		}
	} else {
		a.FreeHost(ap)
	}
}

// AllocInternalP2PPort allocates a unique internal P2P port per VPN container.
// Mirrors Python's alloc_internal_p2p_port.
func (a *PortAllocator) AllocInternalP2PPort(vpn string) int {
	a.mu.Lock()
	defer a.mu.Unlock()
	used := a.vpnP2PUsed[vpn]
	if used == nil {
		used = make(map[int]struct{})
		a.vpnP2PUsed[vpn] = used
	}
	p := 62062
	for {
		if _, taken := used[p]; !taken {
			break
		}
		p++
	}
	used[p] = struct{}{}
	return p
}

func (a *PortAllocator) FreeInternalP2PPort(port int, vpn string) {
	if port == 0 || vpn == "" {
		return
	}
	a.mu.Lock()
	defer a.mu.Unlock()
	if used := a.vpnP2PUsed[vpn]; used != nil {
		delete(used, port)
	}
}

func (a *PortAllocator) ClearAll() {
	a.mu.Lock()
	defer a.mu.Unlock()
	a.usedHost = make(map[int]struct{})
	a.usedHTTP = make(map[int]struct{})
	a.usedHTTPS = make(map[int]struct{})
	a.usedGluetun = make(map[int]struct{})
	a.gluetunNext = a.gluetunMin
	a.vpnPortRanges = make(map[string]*vpnRange)
	a.vpnRangeAssignments = make(map[string]int)
	a.vpnP2PUsed = make(map[string]map[int]struct{})
}

func parseInt(s string) int {
	var n int
	fmt.Sscanf(s, "%d", &n)
	return n
}
